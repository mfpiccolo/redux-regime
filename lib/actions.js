"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dispatchUpdateResourcesByID = exports.normalizeAndMergePayload = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _jsonApiNormalizer = require("json-api-normalizer");

var _jsonApiNormalizer2 = _interopRequireDefault(_jsonApiNormalizer);

var _graphqlNormalizr = require("graphql-normalizr");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var graphQLNormalizr = new _graphqlNormalizr.GraphQLNormalizr();
var graphQlNormalize = graphQLNormalizr.normalize;

var normalizeAndMergePayload = function normalizeAndMergePayload(dispatch, _ref) {
  var jsonApiPayload = _ref.jsonApiPayload,
      graphQlPayload = _ref.graphQlPayload;

  if (jsonApiPayload) {
    Object.entries((0, _jsonApiNormalizer2.default)(jsonApiPayload)).forEach(function (_ref2) {
      var _ref3 = _slicedToArray(_ref2, 2),
          resourceType = _ref3[0],
          resourcesById = _ref3[1];

      dispatch({
        type: "MERGE_RESOURCES",
        resourceType: resourceType,
        resourcesById: resourcesById
      });
    });
  }

  if (graphQlPayload) {
    Object.entries(graphQlNormalize(graphQlPayload)).forEach(function (array) {
      var _array = _slicedToArray(array, 2),
          resourceType = _array[0],
          resourcesById = _array[1];

      dispatch({
        type: "MERGE_RESOURCES",
        resourceType: resourceType,
        resourcesById: _convertToJsonApiSpec(resourceType, resourcesById)
      });
    });
  }
};

var dispatchUpdateResourcesByID = function dispatchUpdateResourcesByID(dispatch, _ref4) {
  var jsonApiPayload = _ref4.jsonApiPayload,
      graphQlPayload = _ref4.graphQlPayload;

  if (jsonApiPayload) {
    _dispatchAddOrReplaceAllJsonApiResources(dispatch, (0, _jsonApiNormalizer2.default)(jsonApiPayload));
  }

  if (graphQlPayload) {
    _dispatchAddOrReplaceAllGraphQlResources(dispatch, "checklists", graphQlNormalize(graphQlPayload));
  }
};

var _dispatchAddOrReplaceAllJsonApiResources = function _dispatchAddOrReplaceAllJsonApiResources(dispatch, resourcesById) {
  Object.entries(resourcesById).forEach(function (_ref5) {
    var _ref6 = _slicedToArray(_ref5, 2),
        resourceType = _ref6[0],
        resources = _ref6[1];

    Object.entries(resources).forEach(function (_ref7) {
      var _ref8 = _slicedToArray(_ref7, 2),
          id = _ref8[0],
          _ref8$ = _ref8[1],
          type = _ref8$.type,
          attributes = _ref8$.attributes,
          links = _ref8$.links,
          relationships = _ref8$.relationships;

      dispatch({
        type: "ADD_OR_REPLACE_RESOURCE_BY_ID",
        resourceType: resourceType,
        id: id,
        attributes: attributes,
        links: links,
        relationships: relationships
      });
    });
  });
};

// TODO: this is not quite done parsing the graphql
var _dispatchAddOrReplaceAllGraphQlResources = function _dispatchAddOrReplaceAllGraphQlResources(dispatch, type, resourcesById) {
  Object.entries(resourcesById).forEach(function (_ref9) {
    var _ref10 = _slicedToArray(_ref9, 2),
        resourceType = _ref10[0],
        resources = _ref10[1];

    Object.entries(resources).forEach(function (_ref11) {
      var _ref12 = _slicedToArray(_ref11, 2),
          id = _ref12[0],
          resource = _ref12[1];

      dispatch({
        type: "ADD_OR_REPLACE_RESOURCE_BY_ID",
        resourceType: resourceType,
        id: id,
        attributes: resource,
        links: null,
        relationships: _buildRelationships(resourceType, resource)
      });
    });
  });
};

var _buildRelationships = function _buildRelationships(type, resource) {
  return Object.entries(resource).reduce(function (newObject, _ref13) {
    var _ref14 = _slicedToArray(_ref13, 2),
        key = _ref14[0],
        value = _ref14[1];

    if (value && Array.isArray(value)) {
      if (!newObject[key]) {
        newObject[key] = { data: [] };
      }

      newObject[key].data = value.map(function (id) {
        return { type: key, id: id };
      });
    }

    if (value && (typeof value === "undefined" ? "undefined" : _typeof(value)) === "object") {
      // TODO: handle hasOne and belongsTo
    }
    return newObject;
  }, {});
};

var _convertToJsonApiSpec = function _convertToJsonApiSpec(resourceType, resourcesById) {
  return Object.entries(resourcesById).reduce(function (formattedResourcesById, _ref15) {
    var _ref16 = _slicedToArray(_ref15, 2),
        id = _ref16[0],
        resource = _ref16[1];

    formattedResourcesById[id] = {
      type: resourceType,
      id: id,
      attributes: resource,
      links: null,
      relationships: _buildRelationships(resourceType, resource)
    };

    return formattedResourcesById;
  }, {});
};

exports.normalizeAndMergePayload = normalizeAndMergePayload;
exports.dispatchUpdateResourcesByID = dispatchUpdateResourcesByID;