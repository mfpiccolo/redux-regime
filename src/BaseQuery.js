import pluralize from "pluralize";

export default class BaseQuery {
  static query(resources) {
    return new QueryObject(pluralize(this.name.toLowerCase()), resources);
  }
}

export class QueryObject {
  constructor(resourceName, resources) {
    this.resourceName = resourceName;
    this.resources = resources;
    this.currentIncludes = [];
    this.currentResources = {};
  }

  all() {
    this._setCurrentResources();
    return this;
  }

  find(id) {
    const resource =
      this.resources[this.resourceName] &&
      this.resources[this.resourceName][id];
    if (resource) {
      this.currentResources = { [id]: resource };
    }
    return this;
  }

  where(params) {
    this._setCurrentResources();
    this._filterAndSetCurrentResourcesByParams(params);
    return this;
  }

  whereRelated(relationship, params) {
    this._setCurrentResources();
    const { resourceName } = this;

    this.currentResources = relationship
      .query(this.resources)
      .where(params)
      .includes([resourceName])
      .execute()
      .reduce((newResource, relatedResource) => {
        const relation = relatedResource[resourceName];
        relation.forEach(({ type, id, ...attributes }) => {
          newResource[id] = { type, id, attributes };
        });
        return newResource;
      }, {});
    return this;
  }

  includes(relationshipTypes) {
    this._setCurrentResources();
    this.currentIncludes = relationshipTypes;
    return this;
  }

  execute() {
    const { currentIncludes, currentResources, _flattenRelationships } = this;
    return Object.values(
      currentResources
    ).map(({ id, attributes, relationships, types, links }) => {
      const newFormattedResource = { id, ...attributes };

      if (!currentIncludes.length) return newFormattedResource;
      return {
        ...newFormattedResource,
        ..._flattenRelationships(
          relationships
        ).reduce((nextRelationshipObjects, { id, type }) => {
          if (!currentIncludes.includes(type)) return nextRelationshipObjects;
          if (!(type in nextRelationshipObjects)) {
            nextRelationshipObjects[type] = [];
          }

          if (!this.resources[type]) return nextRelationshipObjects;
          const relationData = this.resources[type][id];
          if (!relationData) return nextRelationshipObjects;
          nextRelationshipObjects[type].push({
            type,
            id,
            ...relationData.attributes
          });

          return nextRelationshipObjects;
        }, {})
      };
    });
  }

  // Private

  _flattenRelationships(relationships) {
    return Object.values(
      relationships
    ).reduce((nextRelationships, { data }) => {
      return [...nextRelationships, ...data];
    }, []);
  }

  _setCurrentResources() {
    if (this._isEmpty(this.currentResources)) {
      this.currentResources = this.resources[this.resourceName];
    }
  }

  _filterAndSetCurrentResourcesByParams(params) {
    const resourcesByID = Object.entries(
      this.currentResources
    ).reduce((newResource, [id, resource]) => {
      this._filterResourceByParams(params, newResource, resource, id);
      return newResource;
    }, {});
    this.currentResources = resourcesByID;
  }

  _filterResourceByParams(params, newResource, resource, id) {
    Object.entries(params).forEach(([key, value]) => {
      if (resource.attributes[key] === value) {
        newResource[id] = resource;
      }
    });
  }

  _isEmpty(obj) {
    if (
      obj === null ||
      obj === undefined ||
      Array.isArray(obj) ||
      typeof obj !== "object"
    ) {
      return true;
    }
    return Object.getOwnPropertyNames(obj).length === 0 ? true : false;
  }
}
