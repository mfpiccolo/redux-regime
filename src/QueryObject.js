import pluralize from "pluralize";

export default class QueryObject {
  constructor(klass, resourceName, resources, hasMany = [], belongsTo = []) {
    this.klass = klass;
    this.resourceName = resourceName;
    this.resources = resources;
    this.currentIncludes = [];
    this.currentResources = {};
    this.hasMany = hasMany;
    this.belongsTo = belongsTo;
    this._setCurrentResources();
  }

  find(id) {
    const {
      resources,
      resourceName,
      klass,
      _convertToModel,
      hasMany,
      belongsTo
    } = this;
    const { attributes } =
      resources[resourceName] && resources[resourceName][id];
    return _convertToModel(
      klass,
      resources,
      { id, ...attributes },
      hasMany,
      belongsTo
    );
  }

  first() {
    const { resources, resourceName } = this;
    const _resources = resources[resourceName];
    return _resources && _resources[Object.keys(_resources)[0]];
  }

  all() {
    return this;
  }

  where(params) {
    this._filterAndSetCurrentResourcesByParams(params);
    return this;
  }

  whereRelated(relationship, params) {
    const { resourceName } = this;

    this.currentResources = relationship
      .query(this.resources)
      .where(params)
      .includes([resourceName])
      .toObjects()
      .reduce((newResource, relatedResource) => {
        const relation = relatedResource[resourceName];
        relation.forEach(({ type, id, ...attributes }) => {
          newResource[id] = { type, id, attributes };
        });
        return newResource;
      }, {});
    return this;
  }

  includes(relationshipTypes) {
    this.currentIncludes = relationshipTypes;
    return this;
  }

  toModels() {
    return this._reduceCurrentResources("models");
  }

  toObjects() {
    return this._reduceCurrentResources("objects");
  }

  // Private

  _reduceCurrentResources(reducerType) {
    // TODO: needs to be refactored
    const conversion = reducerType === "models"
      ? this._convertToModel
      : this._convertToObject;
    const {
      currentIncludes,
      currentResources,
      resources,
      _flattenRelationships,
      hasMany,
      belongsTo
    } = this;

    return Object.values(
      currentResources
    ).map(({ id, attributes, relationships, types, links }) => {
      const newFormattedResource = conversion(
        this.klass,
        resources,
        {
          id,
          ...attributes
        },
        hasMany,
        belongsTo
      );

      if (!currentIncludes.length) return newFormattedResource;
      return conversion(
        this.klass,
        resources,
        {
          ...newFormattedResource,
          ..._flattenRelationships(
            relationships
          ).reduce((nextRelationshipObjects, { id, type }) => {
            if (!currentIncludes.includes(type)) return nextRelationshipObjects;
            if (!(type in nextRelationshipObjects)) {
              nextRelationshipObjects[type] = [];
            }

            if (!resources[type]) return nextRelationshipObjects;
            const relationData = resources[type][id];
            if (!relationData) return nextRelationshipObjects;
            const relationClass = this.hasMany.find(klass => {
              return pluralize(klass.name.toLowerCase()) === type;
            });

            nextRelationshipObjects[type].push(
              conversion(relationClass, resources, {
                id,
                ...relationData.attributes
              })
            );

            return nextRelationshipObjects;
          }, {})
        },
        hasMany,
        belongsTo
      );
    });
  }

  _convertToModel(klass, resources, resource, hasMany, belongsTo) {
    return new klass(resources, resource, hasMany, belongsTo);
  }

  _convertToObject(klass, resources, resource, hasMany, belongsTo) {
    return resource;
  }

  _flattenRelationships(relationships) {
    return Object.values(
      relationships
    ).reduce((nextRelationships, { data }) => {
      return [...nextRelationships, ...data];
    }, []);
  }

  _setCurrentResources() {
    if (this._isEmpty(this.currentResources) && this.resources) {
      this.currentResources = this.resources[this.resourceName];
    }
  }

  _filterAndSetCurrentResourcesByParams(params) {
    const resourcesByID = Object.entries(
      this.currentResources
    ).reduce((newResource, [id, resource]) => {
      this._filterResourceByParams(params, newResource, resource, id);
      return newResource;
    }, {});
    this.currentResources = resourcesByID;
  }

  _filterResourceByParams(params, newResource, resource, id) {
    Object.entries(params).forEach(([key, value]) => {
      if (key === "id" && resource.id === value) {
        newResource[id] = resource;
      } else if (resource.attributes[key] === value) {
        newResource[id] = resource;
      }
    });
  }

  _isEmpty(obj) {
    if (
      obj === null ||
      obj === undefined ||
      Array.isArray(obj) ||
      typeof obj !== "object"
    ) {
      return true;
    }
    return Object.getOwnPropertyNames(obj).length === 0 ? true : false;
  }
}
