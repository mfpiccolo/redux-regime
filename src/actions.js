import jsonApiNormalize from "json-api-normalizer";
import { GraphQLNormalizr } from "graphql-normalizr";

const graphQLNormalizr = new GraphQLNormalizr();
const graphQlNormalize = graphQLNormalizr.normalize;

const normalizeAndMergePayload = (
  dispatch,
  { jsonApiPayload, graphQlPayload }
) => {
  if (jsonApiPayload) {
    Object.entries(
      jsonApiNormalize(jsonApiPayload)
    ).forEach(([resourceType, resourcesById]) => {
      dispatch({
        type: "MERGE_RESOURCES",
        resourceType,
        resourcesById
      });
    });
  }

  if (graphQlPayload) {
    Object.entries(graphQlNormalize(graphQlPayload)).forEach(array => {
      const [resourceType, resourcesById] = array;
      dispatch({
        type: "MERGE_RESOURCES",
        resourceType,
        resourcesById: _convertToJsonApiSpec(resourceType, resourcesById)
      });
    });
  }
};

const dispatchUpdateResourcesByID = (
  dispatch,
  { jsonApiPayload, graphQlPayload }
) => {
  if (jsonApiPayload) {
    _dispatchAddOrReplaceAllJsonApiResources(
      dispatch,
      jsonApiNormalize(jsonApiPayload)
    );
  }

  if (graphQlPayload) {
    _dispatchAddOrReplaceAllGraphQlResources(
      dispatch,
      "checklists",
      graphQlNormalize(graphQlPayload)
    );
  }
};

const _dispatchAddOrReplaceAllJsonApiResources = (dispatch, resourcesById) => {
  Object.entries(resourcesById).forEach(([resourceType, resources]) => {
    Object.entries(
      resources
    ).forEach(([id, { type, attributes, links, relationships }]) => {
      dispatch({
        type: "ADD_OR_REPLACE_RESOURCE_BY_ID",
        resourceType,
        id,
        attributes,
        links,
        relationships
      });
    });
  });
};

// TODO: this is not quite done parsing the graphql
const _dispatchAddOrReplaceAllGraphQlResources = (
  dispatch,
  type,
  resourcesById
) => {
  Object.entries(resourcesById).forEach(([resourceType, resources]) => {
    Object.entries(resources).forEach(([id, resource]) => {
      dispatch({
        type: "ADD_OR_REPLACE_RESOURCE_BY_ID",
        resourceType,
        id,
        attributes: resource,
        links: null,
        relationships: _buildRelationships(resourceType, resource)
      });
    });
  });
};

const _buildRelationships = (type, resource) => {
  return Object.entries(resource).reduce((newObject, [key, value]) => {
    if (value && Array.isArray(value)) {
      if (!newObject[key]) {
        newObject[key] = { data: [] };
      }

      newObject[key].data = value.map(id => ({ type: key, id }));
    }

    if (value && typeof value === "object") {
      // TODO: handle hasOne and belongsTo
    }
    return newObject;
  }, {});
};

const _convertToJsonApiSpec = (resourceType, resourcesById) => {
  return Object.entries(
    resourcesById
  ).reduce((formattedResourcesById, [id, resource]) => {
    formattedResourcesById[id] = {
      type: resourceType,
      id,
      attributes: resource,
      links: null,
      relationships: _buildRelationships(resourceType, resource)
    };

    return formattedResourcesById;
  }, {});
};

export { normalizeAndMergePayload, dispatchUpdateResourcesByID };
