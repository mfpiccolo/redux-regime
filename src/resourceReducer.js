import merge from "deepmerge";

const initialState = {};

export default function resources(state = initialState, action) {
  const { resourceType } = action;
  switch (action.type) {
    case "ADD_OR_REPLACE_RESOURCE_BY_ID":
      const newState = { ...state };
      const { type, spec, id, attributes, links, relationships } = action;

      _initializeResource(newState, resourceType);

      newState[resourceType][id] = {
        resourceType,
        id,
        attributes,
        links,
        relationships
      };

      return newState;
    case "MERGE_RESOURCES":
      const { resourcesById } = action;
      resources = { ...state };
      if (!resources[resourceType]) {
        resources[resourceType] = {};
      }

      return {
        ...state,
        [resourceType]: merge(resources[resourceType], resourcesById)
      };
    default:
      return state;
  }
}

const _initializeResource = (newState, resourceType) => {
  if (resourceType in newState) return;
  newState[resourceType] = {};
};
