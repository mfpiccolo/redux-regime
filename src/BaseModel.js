import pluralize from "pluralize";
import QueryObject from "./QueryObject";

export default class BaseModel {
  static query(resources) {
    return new QueryObject(
      this,
      pluralize(this.name.toLowerCase()),
      resources,
      this.hasMany,
      this.belongsTo
    );
  }

  constructor(resources, attributes, hasMany = [], belongsTo = []) {
    Object.entries(attributes).forEach(([key, value]) => {
      this[key] = value;
    });

    if (hasMany.forEach) {
      hasMany.forEach(relationship =>
        this._buildHasManyQueryObject(this, resources, relationship)
      );
    }

    if (belongsTo.forEach) {
      belongsTo.forEach(relationship => {
        const relationshipKey = relationship.name.toLowerCase();
        this[relationshipKey] = () => {
          // needs to return the related model
        };
      });
    }
  }

  _filterResources(resource, resources, relationship, relationshipKey) {
    const currentResourceKey = pluralize(
      resource.constructor.name.toLowerCase()
    );
    const resourceClass = resource.constructor;
    const relationshipClass = relationship;
    return {
      ...resources,
      [currentResourceKey]: resources[currentResourceKey][resource.id],
      [relationshipKey]: relationshipClass
        .query(resources)
        .whereRelated(resourceClass, {
          id: resource.id
        }).currentResources
    };
  }

  _buildHasManyQueryObject(resource, resources, relationship) {
    const relationshipKey = pluralize(relationship.name.toLowerCase());
    if (!resource[relationshipKey]) {
      resource[relationshipKey] = () => {
        const newResouces = resource._filterResources(
          resource,
          resources,
          relationship,
          relationshipKey
        );

        return new QueryObject(
          relationship,
          relationshipKey,
          newResouces,
          relationship.hasMany,
          relationship.belongsTo
        );
      };
    }
  }
}
