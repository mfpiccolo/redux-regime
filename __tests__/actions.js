import { normalizeAndMergePayload } from "../src/actions";
import jsonApiPayload
  from "../__testHelpers__/fixtrues/checklistsJsonApiResponse";
import graphQlPayload
  from "../__testHelpers__/fixtrues/checklistsGraphQlResponse";
import normalizedJsonApiTaskPayload
  from "../__testHelpers__/fixtrues/normalizedJsonApiTasksPayload";
import normalizedJsonApiChecklistPayload
  from "../__testHelpers__/fixtrues/normalizedJsonApiChecklistsPayload";
import normalizedGraphQlTaskPayload
  from "../__testHelpers__/fixtrues/normalizedGraphQlTasksPayload";
import normalizedGraphQlChecklistPayload
  from "../__testHelpers__/fixtrues/normalizedGraphQlChecklistsPayload";

describe("actions", () => {
  describe("normalizeAndMergePayload", () => {
    describe("JsonApi", () => {
      test("dispatches MERGE_RESOURCES for each ", () => {
        const dispatch = jest.fn();

        const tasksMergeResourcesAction = {
          resourceType: "tasks",
          resourcesById: normalizedJsonApiTaskPayload,
          type: "MERGE_RESOURCES"
        };

        const checklistsMergeResourcesAction = {
          resourceType: "checklists",
          resourcesById: normalizedJsonApiChecklistPayload,
          type: "MERGE_RESOURCES"
        };

        normalizeAndMergePayload(dispatch, { jsonApiPayload });
        expect(dispatch).toBeCalledWith(tasksMergeResourcesAction);
        expect(dispatch).toBeCalledWith(checklistsMergeResourcesAction);
        expect(dispatch).toMatchSnapshot();
      });
    });

    describe("GraphQl", () => {
      test("dispatches MERGE_RESOURCES for each ", () => {
        const dispatch = jest.fn();

        const tasksMergeResourcesAction = {
          resourceType: "tasks",
          resourcesById: normalizedGraphQlTaskPayload,
          type: "MERGE_RESOURCES"
        };

        const checklistsMergeResourcesAction = {
          resourceType: "checklists",
          resourcesById: normalizedGraphQlChecklistPayload,
          type: "MERGE_RESOURCES"
        };

        normalizeAndMergePayload(dispatch, { graphQlPayload });
        expect(dispatch).toBeCalledWith(tasksMergeResourcesAction);
        expect(dispatch).toBeCalledWith(checklistsMergeResourcesAction);
        expect(dispatch).toMatchSnapshot();
      });
    });
  });
});
