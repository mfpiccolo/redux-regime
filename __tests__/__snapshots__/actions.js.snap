// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions normalizeAndMergePayload GraphQl dispatches MERGE_RESOURCES for each  1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "resourceType": "checklists",
        "resourcesById": Object {
          "4": Object {
            "attributes": Object {
              "id": 4,
              "name": "Onboarding GraphQL",
              "tasks": Array [
                7,
                8,
              ],
            },
            "id": "4",
            "links": null,
            "relationships": Object {
              "tasks": Object {
                "data": Array [
                  Object {
                    "id": 7,
                    "type": "tasks",
                  },
                  Object {
                    "id": 8,
                    "type": "tasks",
                  },
                ],
              },
            },
            "type": "checklists",
          },
          "5": Object {
            "attributes": Object {
              "id": 5,
              "name": "Project Audit GraphQL",
              "tasks": Array [
                9,
                10,
              ],
            },
            "id": "5",
            "links": null,
            "relationships": Object {
              "tasks": Object {
                "data": Array [
                  Object {
                    "id": 9,
                    "type": "tasks",
                  },
                  Object {
                    "id": 10,
                    "type": "tasks",
                  },
                ],
              },
            },
            "type": "checklists",
          },
          "6": Object {
            "attributes": Object {
              "id": 6,
              "name": "QA Checklist GraphQL",
              "tasks": Array [
                11,
                12,
              ],
            },
            "id": "6",
            "links": null,
            "relationships": Object {
              "tasks": Object {
                "data": Array [
                  Object {
                    "id": 11,
                    "type": "tasks",
                  },
                  Object {
                    "id": 12,
                    "type": "tasks",
                  },
                ],
              },
            },
            "type": "checklists",
          },
        },
        "type": "MERGE_RESOURCES",
      },
    ],
    Array [
      Object {
        "resourceType": "tasks",
        "resourcesById": Object {
          "10": Object {
            "attributes": Object {
              "description": "Project Audit Graph 10",
              "id": 10,
            },
            "id": "10",
            "links": null,
            "relationships": Object {},
            "type": "tasks",
          },
          "11": Object {
            "attributes": Object {
              "description": "QA Checklist Graph 11",
              "id": 11,
            },
            "id": "11",
            "links": null,
            "relationships": Object {},
            "type": "tasks",
          },
          "12": Object {
            "attributes": Object {
              "description": "QA Checklist Graph 12",
              "id": 12,
            },
            "id": "12",
            "links": null,
            "relationships": Object {},
            "type": "tasks",
          },
          "7": Object {
            "attributes": Object {
              "description": "Onboarding Graph 7",
              "id": 7,
            },
            "id": "7",
            "links": null,
            "relationships": Object {},
            "type": "tasks",
          },
          "8": Object {
            "attributes": Object {
              "description": "Onboarding Graph 8",
              "id": 8,
            },
            "id": "8",
            "links": null,
            "relationships": Object {},
            "type": "tasks",
          },
          "9": Object {
            "attributes": Object {
              "description": "Project Audit Graph 9",
              "id": 9,
            },
            "id": "9",
            "links": null,
            "relationships": Object {},
            "type": "tasks",
          },
        },
        "type": "MERGE_RESOURCES",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`actions normalizeAndMergePayload JsonApi dispatches MERGE_RESOURCES for each  1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "resourceType": "checklists",
        "resourcesById": Object {
          "1": Object {
            "attributes": Object {
              "name": "Onboarding Rest",
            },
            "id": 1,
            "links": Object {
              "self": "http://example.com/checklists/1",
            },
            "relationships": Object {
              "tasks": Object {
                "data": Array [
                  Object {
                    "id": 1,
                    "type": "tasks",
                  },
                  Object {
                    "id": 2,
                    "type": "tasks",
                  },
                ],
                "links": Object {
                  "related": "http://example.com/checklists/1/tasks",
                  "self": "http://example.com/checklists/1/relationships/tasks",
                },
              },
            },
            "type": "checklists",
          },
          "2": Object {
            "attributes": Object {
              "name": "Project Audit Rest",
            },
            "id": 2,
            "links": Object {
              "self": "http://example.com/checklists/2",
            },
            "relationships": Object {
              "tasks": Object {
                "data": Array [
                  Object {
                    "id": 3,
                    "type": "tasks",
                  },
                  Object {
                    "id": 4,
                    "type": "tasks",
                  },
                ],
                "links": Object {
                  "related": "http://example.com/checklists/2/tasks",
                  "self": "http://example.com/checklists/2/relationships/tasks",
                },
              },
            },
            "type": "checklists",
          },
          "3": Object {
            "attributes": Object {
              "name": "QA Checklist Rest",
            },
            "id": 3,
            "links": Object {
              "self": "http://example.com/checklists/3",
            },
            "relationships": Object {
              "tasks": Object {
                "data": Array [
                  Object {
                    "id": 5,
                    "type": "tasks",
                  },
                  Object {
                    "id": 6,
                    "type": "tasks",
                  },
                ],
                "links": Object {
                  "related": "http://example.com/checklists/3/tasks",
                  "self": "http://example.com/checklists/3/relationships/tasks",
                },
              },
            },
            "type": "checklists",
          },
        },
        "type": "MERGE_RESOURCES",
      },
    ],
    Array [
      Object {
        "resourceType": "tasks",
        "resourcesById": Object {
          "1": Object {
            "attributes": Object {
              "description": "Onboarding REST 1",
            },
            "id": 1,
            "links": Object {
              "self": "http://example.com/tasks/1",
            },
            "type": "tasks",
          },
          "2": Object {
            "attributes": Object {
              "description": "Onboarding REST 2",
            },
            "id": 2,
            "links": Object {
              "self": "http://example.com/tasks/2",
            },
            "type": "tasks",
          },
          "3": Object {
            "attributes": Object {
              "description": "Project Audit Rest 3",
            },
            "id": 3,
            "links": Object {
              "self": "http://example.com/tasks/3",
            },
            "type": "tasks",
          },
          "4": Object {
            "attributes": Object {
              "description": "Project Audit Rest 4",
            },
            "id": 4,
            "links": Object {
              "self": "http://example.com/tasks/4",
            },
            "type": "tasks",
          },
          "5": Object {
            "attributes": Object {
              "description": "QA Checklist Rest 5",
            },
            "id": 5,
            "links": Object {
              "self": "http://example.com/tasks/5",
            },
            "type": "tasks",
          },
          "6": Object {
            "attributes": Object {
              "description": "QA Checklist Rest 6",
            },
            "id": 6,
            "links": Object {
              "self": "http://example.com/tasks/6",
            },
            "type": "tasks",
          },
        },
        "type": "MERGE_RESOURCES",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
